File: backend/app/utils/company.py
import re
import uuid

EMAIL_DOMAIN = "epicnetworkgroup.com"


def _slugify_local_part(text):
    return re.sub(r"[^a-z0-9]", "", (text or "").lower())


def generate_employee_email_base(full_name):
    """first-initial + last-name, e.g. "Jane Doe" -> "jdoe"."""
    parts = [p for p in (full_name or "").strip().split() if p]
    if not parts:
        return "user"
    first = parts[0]
    last = parts[-1] if len(parts) > 1 else ""
    base = _slugify_local_part(first[:1] + last)
    return base or "user"


def generate_department_email_base(name):
    return _slugify_local_part(name) or "dept"


def generate_employee_id_code():
    return f"EIC-{uuid.uuid4().hex[:6].upper()}"


def generate_unique_company_email(full_name):
    from ..models import Employee, User
    base = generate_employee_email_base(full_name)
    candidate = f"{base}@{EMAIL_DOMAIN}"
    suffix = 1
    while User.query.filter_by(email=candidate).first() or Employee.query.filter_by(email=candidate).first():
        suffix += 1
        candidate = f"{base}{suffix}@{EMAIL_DOMAIN}"
    return candidate

File: frontend/src/lib/company.js
export const EMAIL_DOMAIN = "epicnetworkgroup.com";

function slugifyLocalPart(text) {
  return (text || "").toLowerCase().replace(/[^a-z0-9]/g, "");
}

// first-initial + last-name, e.g. "Jane Doe" -> "jdoe"
export function generateEmployeeEmailBase(fullName) {
  const parts = (fullName || "").trim().split(/\s+/).filter(Boolean);
  if (parts.length === 0) return "user";
  const first = parts[0];
  const last = parts.length > 1 ? parts[parts.length - 1] : "";
  return slugifyLocalPart(first[0] + last) || "user";
}

export function generateDepartmentEmailBase(name) {
  return slugifyLocalPart(name) || "dept";
}

// Best-effort client-side uniqueness check against already-loaded records.
// The backend is authoritative (it checks the full database), so this is
// just to avoid an obviously-colliding suggestion in the UI.
export function generateUniqueEmail(base, existingEmails = []) {
  const taken = new Set(existingEmails.filter(Boolean).map((e) => e.toLowerCase()));
  let candidate = `${base}@${EMAIL_DOMAIN}`;
  let suffix = 1;
  while (taken.has(candidate.toLowerCase())) {
    suffix += 1;
    candidate = `${base}${suffix}@${EMAIL_DOMAIN}`;
  }
  return candidate;
}
